### ISSUE TITLE
An error occurs during the validation process.

### ISSUE THREAD

**Issue by LinLin1031** [shenglandu/PushBoundary#5]
![image](https://github.com/shenglandu/PushBoundary/assets/50825626/ca792f4e-e6dd-4290-b07e-200eb9c8de79)

Hello! When running "train_SensatUrban.py", I got an error in the validation phase. Looking forward to your reply. @shenglandu

**Comment by shenglandu** [comment]
Hello,

Can you print truth and bpreds before Line 288 in "trainer_boundary.py"? Another question from my side is that it didn't print training information before validation. Does this mean that the model is not trained but directly validated?

**Comment by LinLin1031** [comment]
> Hello,
> 
> Can you print truth and bpreds before Line 288 in "trainer_boundary.py"? Another question from my side is that it didn't print training information before validation. Does this mean that the model is not trained but directly validated?

![image](https://github.com/shenglandu/PushBoundary/assets/50825626/753fbfbb-8702-45dc-af49-f477f9d0ce21)
![image](https://github.com/shenglandu/PushBoundary/assets/50825626/fc6b3cbd-b491-4a7d-a8bd-6ecae1380721)

Their values are shown in the figure.

![image](https://github.com/shenglandu/PushBoundary/assets/50825626/6d365c44-1af7-49d4-b5a6-623b494df226)

As for the second question, I re-ran the code and the result is as shown.
@shenglandu

**Comment by LinLin1031** [comment]
My problem is with calculating "IoUs for boundary predictions", while there is no error when calculating "IoUs for semantic predictions". @shenglandu 
The problem seems to occur during the validation of the boundary labels, does it mean that the boundary labels of the validation set are incorrect and need to be preprocessed again?

**Comment by LinLin1031** [comment]
> My problem is with calculating "IoUs for boundary predictions", while there is no error when calculating "IoUs for semantic predictions". @shenglandu The problem seems to occur during the validation of the boundary labels, does it mean that the boundary labels of the validation set are incorrect and need to be preprocessed again? 

I just made an attempt to train and validate a few point clouds that I made sure the preprocessing was not a problem. But still the issue occurs. @shenglandu

**Comment by shenglandu** [comment]
> My problem is with calculating "IoUs for boundary predictions", while there is no error when calculating "IoUs for semantic predictions". @shenglandu The problem seems to occur during the validation of the boundary labels, does it mean that the boundary labels of the validation set are incorrect and need to be preprocessed again?

Hello,

Yes, there are errors in the GT boundary labels. In theory, GT true boundary labels should be either 0 or 1, however, from your screenshot, they can range up to 165. The low boundary prediction accuracy (~"30%") during training also shows that your GT boundary label does not match predicted labels.

What is the input to your network (e.g., r, g, b, nx, ny, nz)? Is it possible that you wrongly read a data column such as r g or b as your boundary labels? Otherwise, it's strange that the boundaries can range up to hundreds.


Best Regards,
Shenglan

**Comment by shenglandu** [comment]
> My problem is with calculating "IoUs for boundary predictions", while there is no error when calculating "IoUs for semantic predictions". @shenglandu The problem seems to occur during the validation of the boundary labels, does it mean that the boundary labels of the validation set are incorrect and need to be preprocessed again?

Hello,

Have you solved the issue? 

If not, what I can think of this issue is that for some reason, the boundary labels are wrongly read into the batch. Here are a few locations where you can try to print the boundary labels to check if they are still good (i.e., with a value of only 1 or 0):

- 1) when reading boundaries from files. In "SensatUrban_boundary.py" L398, before "if self.use_potentials", you can print "self.input_boundaries".
- 2) when loading boundaries into the batch. In "SensatUrban_boundary.py" L862, print "self.boundaries".

You can try this and share with me the screenshot. Hope it helps!

Best,
Shenglan

**Comment by LinLin1031** [comment]
@shenglandu Thanks for the heads up. I found out through debugging that the problem was in the process of downsampling the data.
https://github.com/shenglandu/PushBoundary/blob/24e1d965a94c74cb1526517b8bfc4eb1510728b0/KPConv_Backbone/datasets/S3DIS_boundary.py#L348-L351
Specifically, I found that a bug in the above line of code was the source of the problem. (I copied this line of code to the appropriate location in "SensatUrban_boundary.py" to handle SensatUrban.)

Its input features and output sub_features are shown in the figure respectively. Obviously the sub_features are incorrect, but I don't know why at this point.
![Snipaste_2023-09-15_15-45-23](https://github.com/shenglandu/PushBoundary/assets/50825626/e2c9d093-51a0-4abe-9408-1e2587baae1a)
![image](https://github.com/shenglandu/PushBoundary/assets/50825626/6f0df93e-2618-41b9-9d27-0c37781638bd)

**Comment by LinLin1031** [comment]
https://github.com/shenglandu/PushBoundary/blob/24e1d965a94c74cb1526517b8bfc4eb1510728b0/KPConv_Backbone/datasets/S3DIS_boundary.py#L345-L346
One additional detail: since "dis_boundary" is not calculated in "preprocessing_boundaries.py", I'm deleting data['dis_boundary'] from that property, so that's what's causing the problem.

I just tried replacing the position with an attribute that is all zeros and I can get satisfactory results.

**Comment by shenglandu** [comment]
> https://github.com/shenglandu/PushBoundary/blob/24e1d965a94c74cb1526517b8bfc4eb1510728b0/KPConv_Backbone/datasets/S3DIS_boundary.py#L345-L346
> 
> 
> One additional detail: since "dis_boundary" is not calculated in "preprocessing_boundaries.py", I'm deleting data['dis_boundary'] from that property, so that's what's causing the problem.
> I just tried replacing the position with an attribute that is all zeros and I can get satisfactory results.

Hello,

Yes, I think what you observed is correct. To make the subsampling process go smoothly with my data, I modified the original subsampling code (see details [here](https://github.com/shenglandu/PushBoundary/blob/main/KPConv_Backbone/cpp_wrappers/cpp_subsampling/grid_subsampling/grid_subsampling.cpp) L103). It sequentially reads r,g,b, distance, normal, boundary, and direction. If your data doesn't have "dis_boundary", it can mess up the data indices and put the wrong data in the wrong places. "Dis_boundary" was once created but not used in my final experiments. I will update the preprocessing code correspondingly then. 

Best,
Shenglan

**Comment by LinLin1031** [comment]
![image](https://github.com/shenglandu/PushBoundary/assets/50825626/db9c9d4c-29e4-46db-b729-dd05f920b21f)
（1）Now I can run it without any problems. But the above error occurs. @shenglandu 

![image](https://github.com/shenglandu/PushBoundary/assets/50825626/db2bd097-7555-4196-9476-11ed750ef018)
（2）Meanwhile, I trained on a custom dataset similar to SensatUrban. But all 37 epochs have 0.00 IoU, which makes me very puzzled. @shenglandu

**Comment by shenglandu** [comment]
Hello,

In my experiments on S3DIS and SensatUrban, I never experienced "interrupted by 8: sigfpe". Just quickly checked, one possible reason is that the division goes to 0 (check [here](https://github.com/matterport/Mask_RCNN/issues/375) for details). I guess the reason can be that your loss became near 0 (~0.09) which can lead to the vanishing gradient problem, but I'm not sure. In my experiment on SensatUrban, the loss converges to ~0.1 after 500 epochs. So, your loss convergence indicates that overfitting can happen on your dataset.

Here are some tips that might be helpful for your experiment:
- how large is your dataset (i.e., how many tiles did you use for training, validation, and testing)?
- what is the scale of your dataset, and which hyperparameters do you use (e.g., in_radius, sampling grid size)?
- this follows the point above: how many points do you have per batch? If the data distribution is not good, some batches can have a very small number of points, and return a very small loss that is not helpful for the backpropagation.
- are the training labels and validation labels correctly used? (this can be done by printing out the predicted labels and GT labels in each batch, especially in validation).

Good luck with the experiment on your dataset!

Best,
Shenglan

**Comment by Xigui-Li** [comment]
> ![图像](https://private-user-images.githubusercontent.com/50825626/268429278-db9c9d4c-29e4-46db-b729-dd05f920b21f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNDg5MzMsIm5iZiI6MTcwMzA0ODYzMywicGF0aCI6Ii81MDgyNTYyNi8yNjg0MjkyNzgtZGI5YzlkNGMtMjllNC00NmRiLWI3MjktZGQwNWY5MjBiMjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA1MDM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljZDc4NjUwYTA3OWY2Y2M5YzNmMTY1ZDgzNWFlMmVlOGQ0YTViNTIzMTZmYTFmNjRhZmY3NWM3YmMyOTM4MzAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6LEexq6H92aixc3Bp91SsAiW4cZGz4Mg2sXZownGJbM) （1）现在我可以毫无问题地运行它了。但出现上述错误。@shenglandu
> 
> ![图像](https://private-user-images.githubusercontent.com/50825626/268433672-db2bd097-7555-4196-9476-11ed750ef018.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNDg5MzMsIm5iZiI6MTcwMzA0ODYzMywicGF0aCI6Ii81MDgyNTYyNi8yNjg0MzM2NzItZGIyYmQwOTctNzU1NS00MTk2LTk0NzYtMTFlZDc1MGVmMDE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA1MDM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgxZDkxZGYwZmIwMzU5OGFmOWZlZGZjNjkwNjc1NzYzZDc0ODUwMDU5ODcwM2Q0ZjQ1NjcwMTY5MmE2ZjJlZmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3VoShhpqFs7mPaOO0DCmBX1DSArJVz2j-AfkRCaLSYc) （2）同时，我在类似于 SensatUrban 的自定义数据集上进行了训练。但是所有 37 个 epoch 的 IoU 都是 0.00，这让我很困惑。@shenglandu



> ![image](https://private-user-images.githubusercontent.com/50825626/268429278-db9c9d4c-29e4-46db-b729-dd05f920b21f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNDkwMDMsIm5iZiI6MTcwMzA0ODcwMywicGF0aCI6Ii81MDgyNTYyNi8yNjg0MjkyNzgtZGI5YzlkNGMtMjllNC00NmRiLWI3MjktZGQwNWY5MjBiMjFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA1MDUwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNjMWU3MWQ5Y2EwYjJhNjUzNzJiMmQwZmJiOWJjNGRiMDZmZDZmMjVlNWYxYjg4OWQ1M2Q1NzdjNjViNTViOTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rR-A8MEFFRQQE3w0d5igcmkKyjs2y3I4CUMcSG7YI14) （1）Now I can run it without any problems. But the above error occurs. @shenglandu
> 
> ![image](https://private-user-images.githubusercontent.com/50825626/268433672-db2bd097-7555-4196-9476-11ed750ef018.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDMwNDkwMDMsIm5iZiI6MTcwMzA0ODcwMywicGF0aCI6Ii81MDgyNTYyNi8yNjg0MzM2NzItZGIyYmQwOTctNzU1NS00MTk2LTk0NzYtMTFlZDc1MGVmMDE4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjIwVDA1MDUwM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcyZDBkYzdiMGE0Y2I1NGFkZjIxMWRhOGFkMGJjZmNmM2E4M2YzZDBhZTVlZjNiZmQxNWMyNDQwNDhhMmE1MDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6Eb27fGjubNn_6jHIOlb7ZNWOOENDSmrDRc7zVXNXLk) （2）Meanwhile, I trained on a custom dataset similar to SensatUrban. But all 37 epochs have 0.00 IoU, which makes me very puzzled. @shenglandu

Hello, have you solved this problem? I also encountered this issue.

**Comment by LinLin1031** [comment]
> Hello, have you solved this problem? I also encountered this issue.

Sorry, unfortunately, I haven't been able to address these issues either.

**Comment by shenglandu** [comment]
> > Hello, have you solved this problem? I also encountered this issue.
> 
> Sorry, unfortunately, I haven't been able to address these issues either.

Hi,

Have you solved the problem now? If not, can you try the newest code to see if my updates can resolve your issue (0 IoU in validation)?

Best,
Shenglan
